src/differentiation/compute_jacobian_ad.jl

```diff
@@ -41,48 +41,83 @@ function ForwardColorJacCache(f,x,_chunksize = nothing;
+    maxcolor = maximum(colorvec)
+    num_of_chunks = Int(ceil(maxcolor / chunksize))
+    padding_size = (chunksize - (maxcolor % chunksize)) % chunksize
+    partials = colorvec .== (1:maxcolor)'
+    padding_matrix = BitMatrix(undef, length(x), padding_size)
+    partials = hcat(partials, padding_matrix)
 
-    num_of_chunks = Int(ceil(maximum(colorvec) / chunksize))
+    chunked_partials = map(i -> Tuple.(eachrow(partials[:,(i-1)*chunksize+1:i*chunksize])),1:num_of_chunks)
+    chunked_partials
 
-    padding_size = (chunksize - (maximum(colorvec) % chunksize)) % chunksize
+end
 
-    partials = BitMatrix(undef, length(x), maximum(colorvec))
-    partial = BitMatrix(undef, length(x), chunksize)
-    chunked_partials = Array{Array{Tuple{Vararg{Bool,chunksize}},1},1}(
-                                                          undef, num_of_chunks)
+function forwarddiff_color_jacobian(f,
+                x::AbstractArray{<:Number};
+                dx = nothing,
+                colorvec = 1:length(x),
+                sparsity = nothing,
+                jac_prototype = nothing)
+    forwarddiff_color_jacobian(f,x,ForwardColorJacCache(f,x,dx=dx,colorvec=colorvec,sparsity=sparsity),jac_prototype)
+end
 
-    for color_i in 1:maximum(colorvec)
-        for j in 1:length(x)
-            partials[j,color_i] = colorvec[j]==color_i
-        end
-    end
+function forwarddiff_color_jacobian(f,x::AbstractArray{<:Number},jac_cache::ForwardColorJacCache,jac_prototype=nothing)
+    t = jac_cache.t
+    fx = jac_cache.fx
+    dx = jac_cache.dx
+    p = jac_cache.p
+    colorvec = jac_cache.colorvec
+    sparsity = jac_cache.sparsity
+    chunksize = jac_cache.chunksize
+    color_i = 1
+    maxcolor = maximum(colorvec)
 
-    padding_matrix = BitMatrix(undef, length(x), padding_size)
-    partials = hcat(partials, padding_matrix)
+    vecx = vec(x)
+    
+    ncols=length(x)
+    J = jac_prototype isa Nothing ? zeros(Float64,ncols,ncols) : zero(jac_prototype)
 
-    for i in 1:num_of_chunks
-        partial[:,1] .= partials[:,(i-1)*chunksize+1]
-        for j in 2:chunksize
-            partial[:,j] .= partials[:,(i-1)*chunksize+j]
-        end
-        chunked_partials[i] = Tuple.(eachrow(partial))
+    if !(sparsity isa Nothing)
+        rows_index, cols_index = ArrayInterface.findstructralnz(sparsity)
     end
 
-    chunked_partials
-
+    for i in eachindex(p)
+        partial_i = p[i]
+        t = reshape(Dual{typeof(f)}.(vecx, partial_i),size(t))
+        fx = f(t)
+        if !(sparsity isa Nothing)
+            for j in 1:chunksize
+                dx = vec(partials.(fx, j))
+                coords = [(rows_index[i],cols_index[i]) for i in 1:length(cols_index) if colorvec[cols_index[i]] == color_i ]
+                rows_index_c = [coord[1] for coord in coords]
+                cols_index_c = [coord[2] for coord in coords]
+                spmtx = collect(sparse(rows_index_c,cols_index_c,ones(length(rows_index_c)),size(J)...))
```

Comment:
collecting into a sparse matrix will allocate, so we definitely don't want to do that. Should be a non-allocating kernel for static matrices.